browser/services/componentLoader.js

```diff
@@ -0,0 +1,126 @@
+
+const classes = {
+  'virtualbox': VirtualBoxInstall,
+  'cygwin': CygwinInstall,
+  'cdk': CDKInstall,
+  'jdk': JdkInstall,
+  'devstudio': DevstudioInstall,
+  'hyperv': HypervInstall
+};
+
+const baseOrder = {
+  'root': ['jdk'],
+  'jdk':  ['virtualbox', 'devstudio'],
+  'virtualbox': ['hyperv'],
+  'hyperv': ['cygwin'],
+  'cygwin': ['cdk'],
+  'devstudio': [],
+  'cdk': []
+};
+
+class ComponentLoader {
+  constructor(installerDataSvc) {
+    this.requirements = installerDataSvc.requirements;
+    this.installerDataSvc = installerDataSvc;
+  }
+
+  loadComponents() {
+    for (let key in this.requirements) {
+      this.addComponent(key);
+    }
+    this.orderInstallation();
+  }
+
+  loadComponent(key) {
+    this.addComponent(key);
+    this.orderInstallation();
+  }
+
+  removeComponent(key) {
+    this.installerDataSvc.allInstallables().delete(key);
+    this.orderInstallation();
+  }
+
+  addComponent(key) {
+    if (this.requirements[key] && this.requirements[key].bundle !== 'tools') {
+      let skippedProperties = ['name', 'description', 'bundle', 'vendor', 'virusTotalReport'];
+      let folder = key;
+      switch (key) {
+        case 'jdk':
+          folder = 'jdk8'
+          break;
+        case 'devstudio':
+          folder = 'developer-studio';
+          break;
+        case 'hyperv':
+          folder = null;
+          break;
+      }
+      let args = [this.installerDataSvc];
+      if (folder) {
+        args.push(folder);
+      }
+      if (this.requirements[key].dmUrl) {
+        skippedProperties.push('url');
+      }
+
+      for (let property in this.requirements[key]) {
+        if (skippedProperties.indexOf(property) < 0) {
+          args.push(this.requirements[key][property]);
+        }
+      }
+      let component = new DynamicClass(key, args);
+
+      this.installerDataSvc.addItemToInstall(key, component);
+    }
+  }
+
+  orderInstallation() {
+    let newOrder = {};
+    Object.assign(newOrder, baseOrder);
+    let changed;
+
+    do {
+      changed = false;
+      for (let item of Object.keys(newOrder)) {
+        let children = newOrder[item];
+        let finalChildren = [];
+        for (var i = 0; i < children.length; i++) {
+          if (this.installerDataSvc.getInstallable(children[i])) {
+            finalChildren.push(children[i]);
+          } else {
+            if (newOrder[children[i]]) {
+              finalChildren = finalChildren.concat(newOrder[children[i]]);
+              newOrder[item] = finalChildren;
+              changed = true;
+            }
+          }
+        }
+      }
+    } while (changed)
+
+    for (let [key, value] of this.installerDataSvc.allInstallables()) {
+      for (var i = 0; i < newOrder[key].length; i++) {
+        let nextItem = this.installerDataSvc.getInstallable(newOrder[key][i]);
+        value.thenInstall(nextItem);
+      }
+    }
+  }
+}
+
+class DynamicClass {
+  constructor (className, opts) {
+    return new classes[className](...opts);
```

Comment:
This can only be done if the classes define 
```javascript
module.exports = Class
```
and we already have different export mechanism in place so we can use imports. The question is whether it would be beneficial to rewrite exports to module.exports for this.